config/proposer/loader.go

```diff
@@ -0,0 +1,281 @@
+		return nil, err
+	}
+	return ps, nil
+}
+
+func (psl *SettingsLoader) processProposerSettings(loadedSettings, dbSettings *validatorpb.ProposerSettingsPayload) *validatorpb.ProposerSettingsPayload {
+	dbOnly := false
+	if loadedSettings == nil && dbSettings == nil {
+		return nil
+	}
+	// fill in missing data from db
+	if loadedSettings == nil && dbSettings != nil {
+		dbOnly = true
+		loadedSettings = dbSettings
+	}
+	if loadedSettings.DefaultConfig == nil && dbSettings != nil && dbSettings.DefaultConfig != nil {
+		loadedSettings.DefaultConfig = dbSettings.DefaultConfig
+	}
+	if loadedSettings.ProposerConfig == nil && dbSettings != nil && dbSettings.ProposerConfig != nil {
+		loadedSettings.ProposerConfig = dbSettings.ProposerConfig
+	}
+	// if default and proposer configs are both missing even after db setting
+	if loadedSettings.DefaultConfig == nil && loadedSettings.ProposerConfig == nil {
+		return nil
+	}
+
+	if loadedSettings.DefaultConfig != nil {
+		if loadedSettings.DefaultConfig.Builder != nil {
+			loadedSettings.DefaultConfig.Builder.GasLimit = reviewGasLimit(loadedSettings.DefaultConfig.Builder.GasLimit)
+		}
+		// override the db settings with the results based on whether the --enable-builder flag is provided.
+		if psl.options.builderConfig == nil && dbOnly {
+			loadedSettings.DefaultConfig.Builder = nil
+		}
+		if psl.options.builderConfig != nil {
+			o := psl.options.builderConfig.ToConsensus()
+			if loadedSettings.DefaultConfig.Builder != nil {
+				// only override the enabled if builder settings exist
+				loadedSettings.DefaultConfig.Builder.Enabled = o.Enabled
+			} else {
+				loadedSettings.DefaultConfig.Builder = o
+			}
+		} else if psl.options.hasGasLimit && loadedSettings.DefaultConfig.Builder != nil {
+			loadedSettings.DefaultConfig.Builder.GasLimit = psl.options.gasLimit
+		}
+	}
+	if loadedSettings.ProposerConfig != nil && len(loadedSettings.ProposerConfig) != 0 {
+		for _, option := range loadedSettings.ProposerConfig {
+			if option.Builder != nil {
+				option.Builder.GasLimit = reviewGasLimit(option.Builder.GasLimit)
+			}
+			// override the db settings with the results based on whether the --enable-builder flag is provided.
+			if psl.options.builderConfig == nil && dbOnly {
+				option.Builder = nil
+			}
+			if psl.options.builderConfig != nil {
+				o := psl.options.builderConfig.ToConsensus()
+				if option.Builder != nil {
+					// only override the enabled if builder settings exist
+					option.Builder.Enabled = o.Enabled
+				} else {
+					option.Builder = o
+				}
+			} else if psl.options.hasGasLimit && option.Builder != nil {
+				option.Builder.GasLimit = psl.options.gasLimit
+			}
+		}
+	}
+
+	return loadedSettings
+}
+
+func reviewGasLimit(gasLimit validator.Uint64) validator.Uint64 {
+	// sets gas limit to default if not defined or set to 0
+	if gasLimit == 0 {
+		return validator.Uint64(params.BeaconConfig().DefaultBuilderGasLimit)
```

Comment:
Maybe at least add a warning in this case?

Reply:
it might appear too spammy if we add a warning here, by default it should be that default builder gas limit if the builder settings are active.

Reply:
My concern here is that we silently modify user's gas limit without telling him, if they explicitly specifies `0` as a gas limit. Now that `gasLimit` is a pointer, we can differentiate the absence of value and the `0` value.

==> If the value is still `0`, it means that the user explicitly specified it, and we are going to change this value.
(And so in this case IMO it's better to warn the user.)

Reply:
we can add a warning if they explicitly use the flag, but if they don't add gas_limit to the yaml/json file then it'll write the warning even though they didn't explicitly set to 0